packages/network-controller/tests/NetworkController.test.ts

```diff
@@ -3696,6 +3509,143 @@ describe('NetworkController', () => {
       });
     });
   });
+
+  describe('upsertNetworkConfiguration', () => {
+    it('adds the given network configuration when its rpcURL does not match an existing configuration', async () => {
+      (v4 as jest.Mock).mockImplementationOnce(
+        () => 'network-configuration-id-1',
+      );
+      const messenger = buildMessenger();
+      await withController({ messenger }, async ({ controller }) => {
+        const rpcUrlNetwork = {
+          chainId: '0x9999',
+          rpcUrl: 'https://test-rpc.com',
+          ticker: 'RPC',
+        };
+
+        expect(controller.state.networkConfigurations).toStrictEqual({});
+
+        controller.upsertNetworkConfiguration(rpcUrlNetwork, {
+          referrer: 'https://test-dapp.com',
+          source: 'dapp',
+        });
+
+        expect(
+          Object.values(controller.state.networkConfigurations),
+        ).toStrictEqual(
+          expect.arrayContaining([
+            {
+              ...rpcUrlNetwork,
+              nickname: undefined,
+              rpcPrefs: undefined,
+              id: 'network-configuration-id-1',
+            },
+          ]),
+        );
+      });
+    });
+
+    it('update a network configuration when the configuration being added has an rpcURL that matches an existing configuration', async () => {
```

Comment:
@legobeat yeah so this was a forward looking change to serve when `rpcUrl`s are no longer reliably unique. I had [crossed this bit out in the description](https://github.com/MetaMask/core/pull/1064#issue-1531451653) of the PR but it still applies:

> we are considering adding headers (or something like it for credentialed based RPC providers) as an additional part > of these config and so rpcUrls may no longer be a reliably unique field in which case it will make sense for the 
> configs to have a corresponding networkConfigurationId